# Pre-test microbiome train data
<p> First, we use S. Aureus reference sequence as train data.</p>
<p> we download S. Aureus reference sequence from NCBI, after uniformly truncation and coding, model </p>
<p> The same process was then performed for the other seven microbiome species from ZymoBIOMICS HMW DNA Standard D6322. </p>


# Pre-test microbiome test data
<p> we use ZymoBIOMICS HMW DNA Standard as test data.</p>
<p>We sequencing ZymoBIOMICS HMW DNA Standard D6322 with MiniON and get fatsq files. After basecalling, we split sequences as different species. And then we coding them after uniformly truncation with slide windows.</p>






# Human_Large_sacle train data
<p> for the first, download human reference sequence hg38 data from NCBI or other webpage, and then we need to select autosomal and Sex chromosomes.</p>
<p> The complete fasta was trimmed to a length of 200 bp using a sliding window and coded 0123 according to the ACGT nucleotide.</p>


# test data
## how to preprocess data of NA12878 
## for the first, download fastq and fast5 files randomly from NA12878 rel6. for example: 
<p> wget http://s3.amazonaws.com/nanopore-human-wgs/rel6/FASTQTars/FAB42804-84744914_Multi.tar </p>
<p> wget http://s3.amazonaws.com/nanopore-human-wgs/rel6/MultiFast5Tars/FAB42804-84744914_Multi_Fast5.tar </p> 

<p>
Align the downloaded NA12878 data with hg38 to obtain the label, use minimap2.
From sam files we get read id, chrom, and sequuence, select chr1-22, chrM, chrXï¼Œand chrY, delete Unknow and mutation chromosomes.
Then save chromosomes into different txt or fasta files.
</p>


<p> data processing code sample</p>

```bash
## unpack tar files 
tar -xvf FAB42804-84744914_Multi.tar
## unpack .fastq.gz files to fq
gzip -d *.fastq.gz

## install minimap2
conda create -n Minimap2
conda activate Minimap2
conda install -c bioconda minimap2

## install samtools 
conda create -n biotools
conda activate biotools
conda install -c bioconda samtools

## Go to the data folder, use minimap2 to align sequences
minimap2 -ax map-ont hg38/hg38.fa NA12878/Norwich/FAB45280-222619780_Multi/fastq/*.fastq > NA12878/Norwich/FAB45280-222619780_Multi/basecalling_with_hg38/FAB45280-222619780_Multi_alignment.sam

## use samtools and awk to extract information such as read_id, chrom , seq from the alignment results
samtools view -F 2308 FAB46683-4246923067_Multi_alignment.sam | awk '{print ">" $1 , $3 ,$10}'  > FAB46683-4246923067_Multi_all.txt
samtools view -F 2308 FAB46683-4246923067_Multi_alignment.sam |  awk '{print $1 , $3}'  > FAB46683-4246923067_Multi_list.txt

##

```


## Processing of fast5 files (for the comparison part)
```bash
## fast5 files,raw_data is multi style, convert to single
## install  ont-fast5-api
pip install ont-fast5-api

## path to file folder
multi_to_single_fast5 --input_path ./ --save_path fast5 --recursive
```




